### DISCUSSION TITLE
Error Help: Type inside `async` block must be known in this context

### DISCUSSION THREAD

**Discussion by dogou77** [tokio-rs/tokio#5102] [Q&A] [answered]
Hello, I am somewhat of an amateur when it comes to Rust (especially asynchronous programming in it). I am trying to find out a way to see if a task I spawned is still running after a loop so that I can cancel it. The only solution that made sense to me at my skill level was to make a Vector that holds all of the spawned tasks, and then check the vector to see if the tasks are still running. But this led to a bizarre problem that I cannot find a solution for anywhere. I have the following code that is giving me an error:

`let mut tag_log_tasks = Vec::new();`

The error goes as follows:

```
Error[E0698]: type inside `async` block must be known in this context
    --> src/main.rs:100:29
     |
100  |     let mut tag_log_tasks = Vec::new();
     |                             ^^^^^^^^ cannot infer type for type parameter `T` declared on the struct `Vec`
     |
note: the type is part of the `async` block because of this `await`
    --> src/main.rs:1006:139
     |
1006 | ...   let mut func_response = api::update_tags(rack_input.value(), change_input.value(), &global_token, config_data.clone()).await;
     |                                                                                                                             ^^^^^^
```

Line 1006 goes as follows:

```
let mut func_response = api::update_tags(rack_input.value(), change_input.value(), &global_token, config_data.clone()).await;
```

The update_tags function in the api module is defined as follows:

```
pub struct FunctionResponse {
    pub token: String,
    pub error: String,
}

pub async fn update_tags(rack: String, change: String, new_token: &String, config: ConfigData) -> FunctionResponse {
```

My main() function is defined as follows:

```
#[tokio::main]
async fn main() {
```

I cannot post the full code unfortunately, but I was hoping that someone could help give me an idea of what was going on because I am very confused.

**Comment by dogou77** [comment]
I tried to define the type of the Vector with the following code:

`let mut tag_log_tasks: Vec<JoinHandle<_>> = Vec::new();`

But I get the error:

```
error[E0698]: type inside `async` block must be known in this context
    --> src/main.rs:97:9
     |
97   |     let mut tag_log_tasks: Vec<JoinHandle<_>> = Vec::new();
     |         ^^^^^^^^^^^^^^^^^ cannot infer type
     |
note: the type is part of the `async` block because of this `await`
    --> src/main.rs:1004:139
     |
1004 | ...   let mut func_response = api::update_tags(rack_input.value(), change_input.value(), &global_token, config_data.clone()).await;
     |                                                                                                                             ^^^^^^
```

Now I understand that this is because the type inside JoinHandle isn't known (the <_> portion). Does anyone know if there is any type I can put inside there that would fix my issue? The code I am trying to store inside this vec goes as follows:

```
tokio::spawn(async move {
    tokio::time::sleep(Duration::from_secs(5)).await;
    log_buffer_clone.set_text("");
});
```

I appreciate the help.

**Comment by Darksonn** [comment, marked as answer]
It seems like you aren't returning anything from the task, in which case the return type is `()`. In that case, you would use `JoinHandle<()>` as the type.

I note that this error sometimes comes from another error inhibiting type inference. If there are also other errors, you may want to try and fix those first.

You may find [this page](https://tokio.rs/tokio/topics/shutdown) on graceful shutdown useful.

**Reply by dogou77** [reply]
Thanks so much, the page you recommended is really useful.
